### PR TITLE
document how to deal with e-mails

### PR DESCRIPTION
**Pull request by xabbuh**

### CODE REVIEW

File: book/emails.rst [hunk 1]
```diff
@@ -0,0 +1,320 @@
+.. index::
+    :single: E-Mails
+
+Working with E-Mails
+====================
+
+Communicating with customers is an important part of every business. In the Oro Platform, an e-mail
+address is represented by the :class:`Oro\\Bundle\\EmailBundle\\Entity\\Email` activity.
```
Comment (krlove) [inline book/emails.rst]: e-mail address is represented by EmailAddress class, and email instance is represented by Email class

File: book/emails.rst [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+E-Mail Templates
+----------------
+
+When creating e-mails, your users often want to reuse texts they have written some time before for
+a certain purpose. For this, they can create templates and reuse them as needed. Besides letting
+the users create and manage templates through the UI, every bundle can provider their own e-mail
```
Comment (krlove) [inline book/emails.rst]: provider -> provide

File: book/emails.rst [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+templates.
+
+Providing e-mail templates is nothing more than creating Twig templates for the desired format
+(namely HTML or plaintext) in which you can use some pre-defined placeholders to define template
+metadata. Then you create a data fixture that implements the ``AbstractEmailFixture`` class. This
+class provides a ``getEmailsDir`` method which will point to the directory that contains your
```
Comment (krlove) [inline book/emails.rst]: I believe that more accurate phrase would be `getEmailsDir method should return the path of the directory that contains your templates`

File: book/emails.rst [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    // src/Acme/DemoBundle/DataFixtures/ORM/EmailTemplatesFixture.php
+    namespace Acme\DemoBundle\DataFixtures\ORM;
+
+    use Oro\Bundle\EmailBundle\DataFixtures\ORM\AbstractEmailFixture;
+
+    class EmailTemplatesFixture extends AbstractEmailFixture
+    {
+        public function getEmailsDir()
+        {
+            return __DIR__.'/../../Resources/emails';
```
Comment (krlove) [inline book/emails.rst]: This is more preferable implementation of getEmailsDir method's body:
        return $this->container
            ->get('kernel')
            ->locateResource('@AcmeDemoBundle/Migrations/Data/ORM/emails');

File: book/emails.rst [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+E-Mail Address Owners
+---------------------
+
+Each e-mail address is owned by exactly one entity. In the Oro Platform, ``User`` entities and
+``Contact`` entities can be owners of an e-mail address. Supposed you want to use the CRM to keep
+track of all people that apply to your company. You will then probably create an ``Applicant``
+entity and want to associate an e-mail address to each of them. To let your own entities own an
+e-mail address, you have to follow a few steps:
+
+#. :ref:`Create an entity that is response to store the e-mail address. <book-emails-emailinterface>`
```
Comment (krlove) [inline book/emails.rst]: response to story -> responsible for storing

File: book/emails.rst [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+#. :ref:`Create a new owner of an e-mail address. <book-emails-owner>`
+#. :ref:`Publish a provider <book-emails-owner-provider>` that makes it possible to search for th
```
Comment (krlove) [inline book/emails.rst]: th -> the

File: book/emails.rst [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+For your ``Applicant`` entity, the implementation should now look something like this:
+
+.. code-block:: php
+    :linenos:
+
+    // src/Acme/DemoBundle/Entity/Applicant.php
+    namespace Acme\DemoBundle\Entity;
+
+    use Oro\Bundle\EmailBundle\Entity\EmailOwnerInterface;
```
Comment (krlove) [inline book/emails.rst]: Please, add
use Doctrine\ORM\Mapping as ORM;

File: book/emails.rst [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    use Acme\DemoBundle\Entity\Applicant;
+    use Acme\DemoBundle\Entity\ApplicantEmail;
+    use Doctrine\ORM\EntityManager;
+    use Oro\Bundle\EmailBundle\Entity\Provider\EmailOwnerProviderInterface;
+
+    class EmailOwnerProvider implements EmailOwnerProviderInterface
+    {
+        public function getEmailOwnerClass()
+        {
+            return 'Oro\Bundle\UserBundle\Entity\User';
```
Comment (krlove) [inline book/emails.rst]: getEmailOwnerClass should return Applicant entity class:
return 'Acme\DemoBundle\Entity\Applicant';

File: book/emails.rst [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+            if (null === $applicant) {
+                $emailRepository = $em->getRepository('AcmeDemoBundle:Email');
+                $emailEntity = $emailRepository->findOneBy(array('email' => $email));
+
+                if (null !== $emailEntity) {
+                    $applicant = $emailEntity->getEmailOwner();
+                }
+            }
+
+            return $applicant;
```
Comment (krlove) [inline book/emails.rst]: The implementation of findEmailOwner should be changed like so:
        $applicantEmailRepo = $em->getRepository('AcmeDemoBundle:ApplicantEmail');
        $applicantEmail = $applicantEmailRepo->findOneBy(array('email' => $email));

        if (null !== $applicantEmail) {
            return $applicantEmail->getEmailOwner();
        }

        return null;

File: book/emails.rst [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+You then need to create a service for the new ``EmailOwnerProvider`` class and tag it with the
+``oro_email.owner.provider`` tag to make the application aware of the new e-mail provider:
+
+.. code-block:: yaml
+    :linenos:
+
+    # src/Acme/DemoBundle/Resources/config/services.yml
+    services:
+        acme_demo:
```
Comment (krlove) [inline book/emails.rst]: More accurate service name should be specified, for instance:

    acme_demo.provider.email_owner_provider

File: book/emails.rst [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+            class: Acme\DemoBundle\Entity\Provider\EmailOwnerProvider
+            tags:
+                - { name: oro_email.owner.provider, order: 1 }
```
Comment (krlove) [inline book/emails.rst]: Maybe it is better to specify more higher order value, for example `3` - to put newly created provider at the end of providers chain

File: book/emails.rst [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+Refreshing the Database Schema
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Finally, you have to update the database schema and clear the application cache:
+
+.. code-block:: bash
+
+    # update the database schema
+    $ php app/console doctrine:schema:update
```
Comment (krlove) [inline book/emails.rst]: --force flag should be added to execute doctrine:schema:update command

    php app/console doctrine:schema:update --force

**Comment by DimaSoroka** [comment]
@xabbuh please let me know when PR will be updated with review notes

File: book/emails.rst [hunk 13]
```diff
@@ -0,0 +1,340 @@
+.. index::
+    :single: E-Mails
+
+Working with E-Mails
+====================
+
+Communicating with customers is an important part of every business. In the Oro Platform, an e-mail
+address is represented by the :class:`Oro\\Bundle\\EmailBundle\\Entity\\EmailAddress` activity.
```
Comment (akuzmenko-magecore) [inline book/emails.rst]: There is still some confusion:
As was mentioned before - an e-mail address is represented by EmailAddress class, and email instance is represented by Email class. But only Email is **activity**, EmailAddress is not.

Comment (xabbuh) [inline book/emails.rst]: I made another update. Can you check again?

File: book/emails.rst [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+    {
+        /**
+         * @ORM\Id
+         * @ORM\Column(type="integer", name="id")
+         * @ORM\GeneratedValue(strategy="AUTO")
+         */
+        private $id;
+
+        /**
+         * @ORM\OneToMany(targetEntity="Applicant", mappedBy="applicant", orphanRemoval=true, cascade={"persist"})
```
Comment (akuzmenko-magecore) [inline book/emails.rst]: Target entity should be "ApplicantEmail"

File: book/emails.rst [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+The provider class should then look like this:
+
+.. code-block:: php
+    :linenos:
+
+    // src/Acme/DemoBundle/Entity/Provider/EmailOwnerProvider.php
+    namespace Acme\DemoBundle\Entity\Provider;
+
+    use Acme\DemoBundle\Entity\Applicant;
+    use Acme\DemoBundle\Entity\ApplicantEmail;
```
Comment (akuzmenko-magecore) [inline book/emails.rst]: Looks like "Applicant" and "ApplicantEmail" imports are redundant here

**Comment by akuzmenko-magecore** [comment]
Closing this pull request in favor of https://github.com/orocrm/documentation/pull/159
It has been processed manually to resolve conflicts and make review fixes.

**Pull request closed without merging**